### PR TITLE
Remove useless gpu_from_host(host_from_gpu(x)) op.

### PR DESCRIPTION
**Pull request by t13m**
Profiling output after optimization: https://gist.github.com/t13m/f6b2cc1f23229eeabbba#file-profile_after_opt0
Profiling output befoer optimization: https://gist.github.com/t13m/f6b2cc1f23229eeabbba#file-profile_before_opt0

### CODE REVIEW

**Comment by nouiz** [comment]
Can you give us here the optimizer profile before and after this PR?

**Comment by t13m** [comment]
Thanks, the checking is fixed. Host_from_gpu case added. The new profiling output is here: https://gist.github.com/t13m/f6b2cc1f23229eeabbba#file-profile_after_opt0_1
It works, but seems not improve a lot for this profiling example.

I wonder if we can do more gpu_cut_transfer job in as_cuda_ndarray_variable or other opts called before gpu_cut_transfer. Like check_chain(node, gpu_from_host, host_from_gpu).

File: theano/sandbox/cuda/basic_ops.py [hunk 1]
```diff
@@ -34,6 +34,14 @@
 
 
 def as_cuda_ndarray_variable(x):
+    if x.owner and isinstance(x.owner.op, HostFromGpu):
+        return x.owner.inputs[0]
+    if \
+            x.owner and \
```
Comment (nouiz) [inline theano/sandbox/cuda/basic_ops.py]: line 39 and 40 should be merged.

**Comment by nouiz** [comment]
The gpu_cut_transfers opt still remove about 4k nodes. I checked rapidly the file cuda/opt.py and some optimization like local_gpu_elemwise_0 call gpu_from_host instead of as_cuda_ndarray_variable(). This do not use the new optimization in this PR. Can update the opt in that file for this. Do not check the opt in that file before local_gpu_elemwise_0, there use of gpu_from_host is normal.

**Comment by t13m** [comment]
New profiling output: https://gist.github.com/t13m/f6b2cc1f23229eeabbba#file-profile_before_opt0_3

**Comment by nouiz** [comment]
Thanks.

**Comment by t13m** [comment]
Thank you. Backslashes removed in latest commit.

**Pull request merged**